content/lessons/2_ownership/index.md

```diff
@@ -0,0 +1,346 @@
+
+```rust
+let s = String::from("a string"); // 's' is an owner
+
+let s2 = s.clone(); // 's2' now contains its own copy
+
+println!("And the contents are: {}", s); // success!
+```
+
+The compiler is happy now and so are we. The implicit move takes some getting used to, but the compiler is here to help us. Now, let's put the good, old C++ on the table again and compare the two lines:
+
+<div style="text-align: center">
+
+`let s2 = s;` is equivalent to `auto s2 = std::move(s);`
+
+`let s2 = s.clone()` is equivalent to `auto s2 = s`
+
+</div>
+
+There are a few important things to note here:
+
+- Making a copy is oftentimes not cheap. Memory needs to be allocated and copied, and a call to the system has to be made. We should prefer to move things as much as possible to avoid this cost - in C++ we have a myriad of language features like `std::move` and _r-references_ to achieve this. Every programmer worth their salt needs to be well versed in all of them to write efficient C++ code and simply forgetting one move can lead to significant performance loss (and this happens to even the most senior devs ever existing, let's not pretend). On the contrary, in Rust you need to make an effort to make a copy and that makes you very aware of the cost you're paying - something that we'll see quite a lot of in the language. Also, if you forget a clone there's no harm done - it just won't compile!
+
+- Hidden in all of this is another nice thing Rust gives us. In C++, nothing prevents you from using variables after they've been moved from, leading to unexpected errors in a more complex code. In Rust, that variable (in our case `s`) simply becomes invalid and the compiler gives us a nice error about it.
+
+### But what about ints?
+
+A good question to ask. Copying primitives is cheap, cheaper even than moving them. It's not convenient for the programmer either to have to always write `.clone()` after every primitive. If we take a look at the error from the previous example:
+
+```sh
+move occurs because `s` has type `String`, which does not implement the `Copy` trait`
+```
+
+It says that `s` was moved because the `String` type doesn't have the `Copy` trait. We will talk about traits more in depth in the future lessons, but what this basically means is that `String` is not specified to be copied on default. All primitive types (`i32`, `bool`, `f64`, `char`, etc.) and tuples consisting only of primitive types implement the `Copy` trait.
+
+### Exercise
+
+How to fix that code?
+
+```rust
+fn count_animals(nr: u32, animal: String) {
```

Comment:
"num" is the correct English abbreviation for number, not "nr"

Reply:
> also this function doesn't `count` anything, it only prints

The whole example is a reference to counting sheep, hence the name.